app/components/autocomplete/slash_suggestion/app_command_parser/app_command_parser.ts

```diff
@@ -0,0 +1,978 @@
+                    break;
+                }
+                case ' ':
+                case '\t': {
+                    this.state = ParseState.EndCommand;
+                    break;
+                }
+                default:
+                    this.incomplete += c;
+                    this.i++;
+                    break;
+                }
+                break;
+            }
+
+            case ParseState.EndCommand: {
+                const binding = bindings.find((b: AppBinding) => b.label === this.incomplete.toLowerCase());
+                if (!binding) {
+                    // gone as far as we could, this token doesn't match a sub-command.
+                    // return the state from the last matching binding
+                    done = true;
+                    break;
+                }
+                this.binding = binding;
+                this.location += '/' + binding.label;
+                bindings = binding.bindings || [];
+                this.state = ParseState.CommandSeparator;
+                break;
+            }
+
+            case ParseState.CommandSeparator: {
+                if (c === '') {
+                    done = true;
+                }
+
+                switch (c) {
+                case ' ':
+                case '\t': {
+                    this.i++;
+                    break;
+                }
+
+                case '':
+                default: {
+                    this.incomplete = '';
+                    this.incompleteStart = this.i;
+                    this.state = ParseState.Command;
+                    break;
+                }
+                }
+                break;
+            }
+
+            default: {
+                return this.asError('unreachable: unexpected state in matchBinding: ' + this.state);
+            }
+            }
+        }
+
+        if (!this.binding) {
+            return this.asError('"' + this.command + '": no match');
+        }
+
+        this.form = this.binding.form;
+        if (!this.form) {
+            this.form = await this.formsCache.getForm(this.location, this.binding);
+        }
+
+        return this;
+    }
+
+    // parseForm parses the rest of the command using the previously matched form.
+    parseForm = (autocompleteMode = false): ParsedCommand => {
+        if (this.state === ParseState.Error || !this.form) {
+            return this;
+        }
+
+        let fields: AppField[] = [];
+        if (this.form.fields) {
+            fields = this.form.fields;
+        }
+
+        this.state = ParseState.StartParameter;
+        this.i = this.incompleteStart || 0;
+        let flagEqualsUsed = false;
+        let escaped = false;
+
+        // eslint-disable-next-line no-constant-condition
+        while (true) {
+            let c = '';
+            if (this.i < this.command.length) {
+                c = this.command[this.i];
+            }
+
+            switch (this.state) {
+            case ParseState.StartParameter: {
```

Comment:
Nit: Do you think there would be a benefit to creating methods that could handle all the case logic? 

```suggestion
            case ParseState.StartParameter: {
               if(handleStartParameter(c, fields)){
               		return returnedThis;
               }
               break;
            }
            
            ...
            
  func handleStartParameter(c, fields) {
  		switch (c) {
                case '':
                    return this;
                case '-': {
                    // Named parameter (aka Flag). Flag1 consumes the optional second '-'.
                    this.state = ParseState.Flag1;
                    this.i++;
                    return nil;
                }
                default: {
                    // Positional parameter.
                    this.position++;
                    // eslint-disable-next-line no-loop-func
                    const field = fields.find((f: AppField) => f.position === this.position);
                    if (!field) {
                        return this.asError('command does not accept ' + this.position + ' positional arguments');
                    }
                    this.field = field;
                    this.state = ParseState.StartValue;
                    return nil;
                }
         }                        
            
```

Reply:
I'm not sure, it may make things more readable. @levb do you have thoughts on this?

Reply:
I prefer it in one large function, easier for me to read and iterate. What needs to be consolidated into little functions are state transitions and `incomplete` management.